src/common/meta/src/ddl/create_tables.rs

```diff
@@ -0,0 +1,381 @@
+            let request_builder = self.create_region_request_builder(physical_table_id, task)?;
+
+            for region_number in &regions {
+                let region_id = RegionId::new(logical_table_id, *region_number);
+                let create_region_request = request_builder.build_one(
+                    region_id,
+                    storage_path.clone(),
+                    region_wal_options,
+                )?;
+                requests.push(create_region_request);
+            }
+        }
+
+        Ok(CreateRequests { requests })
+    }
+
+    async fn create_regions(&mut self, region_routes: &[RegionRoute]) -> Result<Status> {
+        let leaders = find_leaders(region_routes);
+        let mut create_region_tasks = Vec::with_capacity(leaders.len());
+
+        for datanode in leaders {
+            let requester = self.context.datanode_manager.datanode(&datanode).await;
+            let creates = self.one_datanode_region_requests(&datanode, region_routes)?;
+            let request = RegionRequest {
+                header: Some(RegionRequestHeader {
+                    tracing_context: TracingContext::from_current_span().to_w3c(),
+                    ..Default::default()
+                }),
+                body: Some(PbRegionRequest::Creates(creates)),
+            };
+            create_region_tasks.push(async move {
+                if let Err(err) = requester.handle(request).await {
+                    return Err(handle_operate_region_error(datanode)(err));
+                }
+                Ok(())
+            });
+        }
+
+        join_all(create_region_tasks)
+            .await
+            .into_iter()
+            .collect::<Result<Vec<_>>>()?;
+
+        self.creator.data.state = CreateTablesState::CreateMetadata;
+
+        // Ensures the procedures after the crash start from the `DatanodeCreateRegions` stage.
+        Ok(Status::executing(false))
+    }
+}
+
+#[async_trait]
+impl Procedure for CreateLogicalTablesProcedure {
+    fn type_name(&self) -> &str {
+        Self::TYPE_NAME
+    }
+
+    async fn execute(&mut self, _ctx: &ProcedureContext) -> ProcedureResult<Status> {
+        let state = &self.creator.data.state;
+
+        let _timer = metrics::METRIC_META_PROCEDURE_CREATE_TABLES
+            .with_label_values(&[state.as_ref()])
+            .start_timer();
+
+        match state {
+            CreateTablesState::Prepare => self.on_prepare().await,
+            CreateTablesState::DatanodeCreateRegions => self.on_datanode_create_regions().await,
+            CreateTablesState::CreateMetadata => self.on_create_metadata().await,
+        }
+        .map_err(handle_retry_error)
+    }
+
+    fn dump(&self) -> ProcedureResult<String> {
+        serde_json::to_string(&self.creator.data).context(ToJsonSnafu)
+    }
+
+    fn lock_key(&self) -> LockKey {
+        let table_id = self.creator.data.physical_table_id();
+        LockKey::single(TableLock::Write(table_id))
+    }
```

Comment:
Needs to acquire locks(`TableNameLock`) of all creating table names

Reply:
‌‌Just lock the physical table is ok, and it will check if the logical tables exists inside the lock.

Reply:
You are right,  it is necessary to lock the table name because there will be competition with creating physical tables.